src/typing_extensions.py

```diff
@@ -352,27 +352,27 @@ def clear_overloads():
 TYPE_CHECKING = typing.TYPE_CHECKING
 
 
-_PROTO_WHITELIST = ['Callable', 'Awaitable',
+_PROTO_WHITELIST = {'Callable', 'Awaitable',
                     'Iterable', 'Iterator', 'AsyncIterable', 'AsyncIterator',
                     'Hashable', 'Sized', 'Container', 'Collection', 'Reversible',
-                    'ContextManager', 'AsyncContextManager']
+                    'ContextManager', 'AsyncContextManager'}
 
 
 def _get_protocol_attrs(cls):
     attrs = set()
     for base in cls.__mro__[:-1]:  # without object
-        if base.__name__ in ('Protocol', 'Generic'):
+        if base.__name__ in {'Protocol', 'Generic'}:
             continue
         annotations = getattr(base, '__annotations__', {})
-        for attr in list(base.__dict__.keys()) + list(annotations.keys()):
-            if (not attr.startswith('_abc_') and attr not in (
+        for attr in (*base.__dict__.keys(), *annotations.keys()):
```

Comment:
How much faster is `*`-unpacking here compared to list addition? Can you show some benchmarks?

Reply:
Given that dictionary keys are set-like objects, the following is probably more idiomatic, and might be faster yet:

```suggestion
        for attr in base.__dict__.keys() | annotations.keys():
```

Reply:
There are benchmarks for `list(d1.keys()) + list(d2.keys())` (1), `(*d1.keys(), *d2.keys())` (2) and `d1.keys() | d2.keys()` (3).
I am using CPython 3.10.4.

For small dicts: (2) is 20-30% faster than (1), (3) is 12% faster than (1) if we dont iterate over result and is 11% slower than (1) if we do iterate over result.
For huge dicts: (2) is 5% than (1), (3) is much slower than (1).

So I prefer unpacking into tuple over list addition.

```py
Small dicts:
$>py -m timeit -s "d1 = {'a': 1, 'b': 2, 'c': 3}; d2 = {'x': 4, 'y': 5, 'z': 6}" "list(d1.keys()) + list(d2.keys())"
1000000 loops, best of 5: 315 nsec per loop

$>py -m timeit -s "d1 = {'a': 1, 'b': 2, 'c': 3}; d2 = {'x': 4, 'y': 5, 'z': 6}" "(*d1.keys(), *d2.keys())"
1000000 loops, best of 5: 218 nsec per loop

$>py -m timeit -s "d1 = {'a': 1, 'b': 2, 'c': 3}; d2 = {'x': 4, 'y': 5, 'z': 6}" "d1.keys() | d2.keys()"
1000000 loops, best of 5: 276 nsec per loop

Also measure time for iterating over result:
$>py -m timeit -s "d1 = {'a': 1, 'b': 2, 'c': 3}; d2 = {'x': 4, 'y': 5, 'z': 6}" "list(list(d1.keys()) + list(d2.keys()))"
1000000 loops, best of 5: 372 nsec per loop

$>py -m timeit -s "d1 = {'a': 1, 'b': 2, 'c': 3}; d2 = {'x': 4, 'y': 5, 'z': 6}" "list((*d1.keys(), *d2.keys()))"
1000000 loops, best of 5: 296 nsec per loop

$>py -m timeit -s "d1 = {'a': 1, 'b': 2, 'c': 3}; d2 = {'x': 4, 'y': 5, 'z': 6}" "list(d1.keys() | d2.keys())"
500000 loops, best of 5: 413 nsec per loop



Large dicts:
$>py -m timeit -s "d1 = dict.fromkeys(range(10000)); d2 = dict.fromkeys(range(10000,20000))" "list(d1.keys()) + list(d2.keys())"
1000 loops, best of 5: 165 usec per loop

$>py -m timeit -s "d1 = dict.fromkeys(range(10000)); d2 = dict.fromkeys(range(10000,20000))" "(*d1.keys(), *d2.keys())"
1000 loops, best of 5: 159 usec per loop

$>py -m timeit -s "d1 = dict.fromkeys(range(10000)); d2 = dict.fromkeys(range(10000,20000))" "d1.keys() | d2.keys()"
500 loops, best of 5: 720 usec per loop

Also measure time for iterating over result:
$>py -m timeit -s "d1 = dict.fromkeys(range(10000)); d2 = dict.fromkeys(range(10000,20000))" "list(list(d1.keys()) + list(d2.keys()))"
1000 loops, best of 5: 211 usec per loop

$>py -m timeit -s "d1 = dict.fromkeys(range(10000)); d2 = dict.fromkeys(range(10000,20000))" "list((*d1.keys(), *d2.keys()))"
1000 loops, best of 5: 202 usec per loop

$>py -m timeit -s "d1 = dict.fromkeys(range(10000)); d2 = dict.fromkeys(range(10000,20000))" "list(d1.keys() | d2.keys())"
500 loops, best of 5: 821 usec per loop
```

Reply:
Thanks, that's a convincing benchmark! My suggestion definitely loses the speed race, in that case.

Looks like the fastest way yet, however, might be to drop the call to `.keys()`, which isn't actually necessary here at all:

```
$>py -m timeit -s "d1 = {'a': 1, 'b': 2, 'c': 3}; d2 = {'x': 4, 'y': 5, 'z': 6}" "list(list(d1.keys()) + list(d2.keys()))"
1000000 loops, best of 5: 364 nsec per loop

$>py -m timeit -s "d1 = {'a': 1, 'b': 2, 'c': 3}; d2 = {'x': 4, 'y': 5, 'z': 6}" "list((*d1.keys(), *d2.keys()))"
1000000 loops, best of 5: 296 nsec per loop

$>py -m timeit -s "d1 = {'a': 1, 'b': 2, 'c': 3}; d2 = {'x': 4, 'y': 5, 'z': 6}" "list(d1.keys() | d2.keys())"
500000 loops, best of 5: 415 nsec per loop

$>py -m timeit -s "d1 = {'a': 1, 'b': 2, 'c': 3}; d2 = {'x': 4, 'y': 5, 'z': 6}" "list((*d1, *d2))"
1000000 loops, best of 5: 208 nsec per loop
```

```suggestion
        for attr in (*base.__dict__, *annotations):
```

Reply:
Didn't notice it :)